From 479227ac6686baa0e662ac8c130ed961395710cd Mon Sep 17 00:00:00 2001
From: Jianqun Xu <jay.xu@rock-chips.com>
Date: Wed, 16 Dec 2020 19:39:05 +0800
Subject: [PATCH] android: ion: add debug node for ion heap

Add debug node [system/cma]_stats for ion [system/cma]-heap

Change-Id: Ib8ea98f516586c9598e17efe310f7673421fa02a
Signed-off-by: Jianqun Xu <jay.xu@rock-chips.com>
---
 drivers/staging/android/ion/ion.c | 30 ++++++++++++++++++++++++++++++
 1 file changed, 30 insertions(+)

diff --git a/drivers/staging/android/ion/ion.c b/drivers/staging/android/ion/ion.c
index a1721b0bddf0..87bc7ffaf6f6 100644
--- a/drivers/staging/android/ion/ion.c
+++ b/drivers/staging/android/ion/ion.c
@@ -570,6 +570,28 @@ static const struct file_operations ion_fops = {
 #endif
 };
 
+static int ion_debug_heap_show(struct seq_file *s, void *unused)
+{
+	struct ion_heap *heap = s->private;
+
+	if (heap->debug_show)
+		heap->debug_show(heap, s, unused);
+
+	return 0;
+}
+
+static int ion_debug_heap_open(struct inode *inode, struct file *file)
+{
+	return single_open(file, ion_debug_heap_show, inode->i_private);
+}
+
+static const struct file_operations debug_heap_fops = {
+	.open = ion_debug_heap_open,
+	.read = seq_read,
+	.llseek = seq_lseek,
+	.release = single_release,
+};
+
 static int debug_shrink_set(void *data, u64 val)
 {
 	struct ion_heap *heap = data;
@@ -644,6 +666,14 @@ void ion_device_add_heap(struct ion_heap *heap)
 				    heap, &debug_shrink_fops);
 	}
 
+	if (heap->debug_show) {
+		char debug_name[64];
+
+		snprintf(debug_name, 64, "%s_stats", heap->name);
+		debugfs_create_file(debug_name, 0644, dev->debug_root,
+				    heap, &debug_heap_fops);
+	}
+
 	dev->heap_cnt++;
 	up_write(&dev->lock);
 }
-- 
2.35.3

